DM-44842: Fix query crashes on postgres

dhirving opened the pull request:
Fixed two issues where invalid SQL statements were being generated for Postgres by the new query system.  (See individual commit messages for details.)  

We now also run the Butler server tests using Postgres for the registry, since this is the configuration being used in production.  In the process, deduplicated the logic for setting up a postgres instance for testing -- it had been copy/pasted five places.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Fix query crash on postgres

Fix an issue where exceptions like 'psycopg2.errors.GroupingError: column "patch.region" must appear in the GROUP BY clause or be used in an aggregate function' would be thrown for some queries.  

This would occur for queries with spatial constraints on dimensions that are not returned by the query.

dhirving committed:
Deduplicate postgres testing logic

The same postgres testing setup code had been copy/pasted to five different locations, and I'm about to need it again somewhere else.  Factor out this logic to a reusable class.

dhirving committed:
Pin numpy < 2

A new major release of numpy just came out and we're not compatible with it yet.

dhirving committed:
Fix postgres query.datasets tagged collection crash

Fix a bug where dataset queries involving tagged collections would fail to run on Postgres when using the SynthIntKeyCollectionManager.  These queries would fail with an exception like "sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedTable) missing FROM-clause entry for table "dataset"
E       LINE 2: ...space_cfd3f837d0f16f91.collection.collection_id = namespace_..."

This was occurring because this case joins both the collection and static dataset tables.  The ON clause when joining the collection table references the static dataset table, but we weren't joining the dataset table until after the collection table.

dhirving committed:
Run server registry tests with a postgres backend

The normal deployment scenario for Butler server uses Postgres, and many tests for the new query system currently only run through RemoteButler.  So to help catch postgres issues earlier, run the unit test suite against the server backed by both postgres and sqlite.

dhirving committed:
Run main test suite against server with postgres

In addition to the registry tests in the previous commit, run the main test suite against Butler server with a postgres backend to ensure compatibility.

dhirving committed:
Pull out reusable postgres version function

TallJimbo commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -686,6 +688,11 @@ def apply_query_projection(self, plan: QueryProjectionPlan, builder: QueryBuilde
+        # derived fields, since there is only one region associated with each
+        # dimension key value.
+        for element in plan.region_non_aggregates:
+            derived_fields.append((element.name, "region"))
```
I think I originally expected these to be added by the loop below, since regions are a kind of dimension field.  The bug was that that was only happening for regions included in the output rows, not regions needed only by `Postprocessing`, since we don't put the latter in `plan.columns`.  I think that means we should also guard against a region appearing in both places.

To do that might make sense to drop the `region_aggregate` and `region_non_aggregate` attributes from `QueryProjectionPlan`, and instead handle them both together here in one loop:
```
for element in builder.postprocessing.iter_missing(plan.columns):
    if element.name not in plan.columns.dimensions.elements:
        builder.joiner.fields[element.name]["region"] = ddl.Base64Region.union_aggregate(
            builder.joiner.fields[element.name]["region"]
        )
        have_aggregates = True
    else:
        derived_fields.append((element.name, "region"))
```
It doesn't look like we need to split out the aggregate regions from the non-aggregate regions in the plan data structure anywhere else.

TallJimbo commented on python/lsst/daf/butler/tests/butler_queries.py:
```diff
@@ -467,6 +467,21 @@ def test_common_skypix_overlaps(self) -> None:
+                    (data_id["tract"], data_id["patch"])
+                    for data_id in query.data_ids(["patch"]).where({"instrument": "HSC", "visit": 318})
+                ],
+            )
```
A version of this test that returned patch dimension records instead of data coordinates would cover the case of the region appear in both the result rows and the `Postprocessing` that I was worried about in my other comment, but I think the original query system tests I wrote should have covered that already. I'm guessing that means the duplication I was worried about either gets fixed already in another layer or it manifests as something harmless, like a column nothing uses in the SELECT clause.

dhirving replied:
Added the test and it appears fine.

Unless I'm missing something there shouldn't be any duplication, though -- isn't the point of `iter_missing` that it doesn't include columns that already appear in the result rows?

TallJimbo replied:
Right, with the `iter_missing` version there shouldn't be any duplication.  I think with the previous version of this PR there would have been duplicate entries in `derived_fields`, but I don't know if that would have caused any real problems.

TallJimbo commented on tests/test_postgresql.py:
```diff
@@ -27,75 +27,45 @@
-        cls.server.stop()
-        removeTestTempDir(cls.root)
+        super().setUpClass()
+        cls.postgres = cls.enterClassContext(setup_postgres_test_db())
```
I was getting all ready to be concerned about using a context manager for this utility code since I knew it was going to be needed in a `setUpClass`, but I didn't know about `enterClassContext`.  Nice!

dhirving committed:
Simplify projection query region logic

Instead of having the logic for region fields span multiple functions with variables in between, handle it all in one place.
